feat: Employee Hours Utilization Report

NagariaHussain opened the pull request:
<!--

Some key notes before you open a PR:

 1. Select which branch should this PR be merged in?
 2. PR name follows [convention](http://karma-runner.github.io/4.0/dev/git-commit-msg.html)
 3. All tests pass locally, UI and Unit tests
 4. All business logic and validations must be on the server-side
 5. Update necessary Documentation
 6. Put `closes #XXXX` in your comment to auto-close the issue that your PR fixes


Also, if you're new here

- Documentation Guidelines => https://github.com/frappe/erpnext/wiki/Updating-Documentation

- Contribution Guide => https://github.com/frappe/erpnext/blob/develop/.github/CONTRIBUTING.md

- Pull Request Checklist => https://github.com/frappe/erpnext/wiki/Pull-Request-Checklist

-->

**This report gives you an overview of the utilization of employee working hours based on Timesheets submitted.**


![projects-employee-util-report](https://user-images.githubusercontent.com/34810212/116075083-24fdcd80-a6b0-11eb-9a6c-7aafc4835e44.png)


<!-- You can skip this if you're fixing a typo or updating existing documentation -->

The Employee Working hours are divided into 3 categories:
    
1. Billed Hours
2. Non-Billed Hours
3. Un-tracked Hours (Total - Billed - Non-Billed)

The percentage utilization for each employee is calculated using the below formula:

```
% Utilization = [(Billed Hours + Non-Billed Hours) / (Total Hours)] * 100
```

where **Total Hours** is the number of working hours in the selected *date range*.

The report can also be filtered for a particular Employee, Department and/or Project.

<!-- Example: When "Adding a function to do X", explain why it is necessary to have a way to do X. -->

![report2](https://user-images.githubusercontent.com/34810212/115011254-0dfaf680-9ecc-11eb-93c7-365d51b20531.png)

![report3](https://user-images.githubusercontent.com/34810212/115011269-13584100-9ecc-11eb-9103-d1b78adbe225.png)


<!-- Add images/recordings to better visualize the change: expected/current behviour -->

**Documentation PR**: https://github.com/frappe/erpnext_documentation/pull/281

NagariaHussain committed:
feat: Add Report boilerplate

NagariaHussain committed:
feat: Add filters to front end

NagariaHussain committed:
feat: Add report table data generation

NagariaHussain committed:
feat: Add Report Summary generation

NagariaHussain committed:
fix: Missing quotation marks in SQL query

NagariaHussain committed:
feat: Add chart generation

NagariaHussain committed:
test: Basic report data testing

NagariaHussain committed:
feat: Add additional stats to report summary

NagariaHussain committed:
test: Add test case for summary data

NagariaHussain committed:
fix: Handle overtime edge case

NagariaHussain committed:
fix: Rename utilisation to utilization

NagariaHussain committed:
test: Unique Employee IDs for tests

NagariaHussain committed:
fix: Sider Issues

nabinhait committed:
Merge branch 'develop' into employee-util-report

ruchamahabal commented on erpnext/projects/report/employee_hours_utilization_based_on_timesheet/employee_hours_utilization_based_on_timesheet.py:
```diff
@@ -0,0 +1,228 @@
+				ON ttd.parent = tt.name
+			WHERE tt.start_date BETWEEN '{0}' AND '{1}'
+			AND tt.end_date BETWEEN '{0}' AND '{1}'
+			{2};  
```
Only fetch records where Employee is linked since the Employee Link field is optional in Timesheets:

![image](https://user-images.githubusercontent.com/24353136/114822688-1249e580-9de0-11eb-8f96-44689110e4b3.png)

ruchamahabal commented on erpnext/projects/report/employee_hours_utilization_based_on_timesheet/employee_hours_utilization_based_on_timesheet.py:
```diff
@@ -0,0 +1,228 @@
+				'options': 'Employee',
+				'fieldname': 'employee',
+				'fieldtype': 'Link',
+				'width': 200
```
Add 2 more column
- Employee Name
- Department

Add Department column to filters too.

ruchamahabal commented on erpnext/projects/report/employee_hours_utilization_based_on_timesheet/employee_hours_utilization_based_on_timesheet.py:
```diff
@@ -0,0 +1,228 @@
+				'width': 150
+			},
+			{
+				'label': _('% Utilization'),
```
Let's also show the formula here

ruchamahabal commented on erpnext/projects/report/employee_hours_utilization_based_on_timesheet/employee_hours_utilization_based_on_timesheet.py:
```diff
@@ -0,0 +1,228 @@
+		additional_filters = ''
+
+		if self.filters.employee:
+			additional_filters += f"AND tt.employee = '{self.filters.employee}'"
```
+1 for using f-strings.

ruchamahabal reviewed (commented):
Congratulations on your first contribution 🥳 . Nice PR 👌
Suggesting some changes.

NagariaHussain committed:
feat: Add filtering by department

Also:

- Add department filter to js
- Add department column to report
- Fetch only those timesheets which have an Employee Linked
- Update unit tests

NagariaHussain committed:
Merge branch 'employee-util-report' of https://github.com/NagariaHussain/erpnext into employee-util-report

ruchamahabal commented:
Add the Report to Project Workspace

NagariaHussain committed:
chore: Add Report to Dashboard

NagariaHussain commented:
> Add the Report to Project Workspace

Done.

ruchamahabal committed:
Merge branch 'develop' into employee-util-report

ruchamahabal committed:
Merge branch 'develop' into employee-util-report

ruchamahabal committed:
fix: validate Standard Working Hours

- use standard working hours in metrics calculation

ruchamahabal committed:
chore: clean-up

- set default date filters for last month instead of year

- added abbreviations for columns to use them in the metrics labels

ruchamahabal committed:
feat: Add % Utilization (Billed Only)

- Added billed utilization metric in report summary

ruchamahabal committed:
fix: tests

ruchamahabal committed:
Merge branch 'develop' into employee-util-report

ruchamahabal committed:
fix: add employee name in report data

ruchamahabal commented:
All tests are passing but submitting the coverage is failing which is unrelated to this PR, hence merging.

![test](https://user-images.githubusercontent.com/24353136/116076007-4dd29280-a6b1-11eb-901b-6c78689280f4.png)
